{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fasta header: TestSeq\n",
      "\n",
      "AT repeat discovered 1 time(s)\n",
      "\n",
      "AT repeat first found at positions 89 - 105\n",
      "AT repeat discovered 1 time(s) across 1 records.\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Seq_ID</th>\n",
       "      <th>descrption</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M91373; SV 1; linear; mRNA; STD; PLN; 1131 BP.</td>\n",
       "      <td>Cucumis sativus peroxidase mRNA, complete cds.</td>\n",
       "      <td>ACCAGAGAAGACCCCATTTGCAGTATCAAAAATGGGTTTACCTAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M57705; SV 1; linear; mRNA; STD; ROD; 237 BP.</td>\n",
       "      <td>Rat truncated thyroid peroxidase mRNA, 3' end.</td>\n",
       "      <td>CATCGATCATGACATTGCTCTCACACCACAGAGCACCAGCACAGCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                           Seq_ID  \\\n",
       "0  M91373; SV 1; linear; mRNA; STD; PLN; 1131 BP.   \n",
       "1   M57705; SV 1; linear; mRNA; STD; ROD; 237 BP.   \n",
       "\n",
       "                                       descrption  \\\n",
       "0  Cucumis sativus peroxidase mRNA, complete cds.   \n",
       "1  Rat truncated thyroid peroxidase mRNA, 3' end.   \n",
       "\n",
       "                                            sequence  \n",
       "0  ACCAGAGAAGACCCCATTTGCAGTATCAAAAATGGGTTTACCTAAA...  \n",
       "1  CATCGATCATGACATTGCTCTCACACCACAGAGCACCAGCACAGCA...  "
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import re\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "def load_fasta(fasta_path):\n",
    "    \"\"\"Without Using Biopython\"\"\"\n",
    "    fasta = open(fasta_path, 'r').read()\n",
    "    record_dict = dict()\n",
    "    records = fasta.split('>')\n",
    "    for rec in records:\n",
    "        if rec:\n",
    "            rec_lines = rec.split('\\n')\n",
    "            header = rec_lines[0]\n",
    "            seq = ''.join(rec_lines[1:]).replace(' ', '')\n",
    "            record_dict[header] = seq\n",
    "    return record_dict\n",
    "\n",
    "def check_seq_for_motif(seq):\n",
    "    \"\"\"\n",
    "    Check seq for motif\n",
    "    Input:\n",
    "    seq: str, sequence\n",
    "    motif: str, motif to search seq for\n",
    "    \"\"\"\n",
    "    pattern = '(AT){3,}'\n",
    "    r_all = re.findall(pattern, seq)\n",
    "    r_first = re.search(pattern, seq)\n",
    "    at_count = len(r_all)\n",
    "    if r_first:\n",
    "        start, end = r_first.span()\n",
    "        print(f'AT repeat discovered {at_count} time(s)\\n')\n",
    "        print(f'AT repeat first found at positions {start} - {end}')\n",
    "\n",
    "    else:\n",
    "        print(f'Unable to find AT repeat\\n\\n')\n",
    "        \n",
    "    return at_count\n",
    "        \n",
    "        \n",
    "def check_all_seqs(fasta_path):\n",
    "    \"\"\"\n",
    "    Input:\n",
    "    fasta_path: str, path to fasta file\n",
    "    motif: str, motif to search for within fasta\n",
    "    \n",
    "    \"\"\"\n",
    "    if os.path.exists(fasta_path):\n",
    "        record_dict = load_fasta(fasta_path)\n",
    "    else:\n",
    "        print(f'Fasta path {fasta_path} does not exist.')\n",
    "        raise OSError\n",
    "    total_count = 0\n",
    "    N_records = 0\n",
    "    for header in record_dict:\n",
    "        N_records += 1\n",
    "        print(f'Fasta header: {header}\\n')\n",
    "        tcount = check_seq_for_motif(record_dict[header])\n",
    "        total_count += tcount\n",
    "    print(f'AT repeat discovered {total_count} time(s) across {N_records} records.\\n\\n\\n')\n",
    "\n",
    "    \n",
    "def parse_embl(embl_path, write=True):\n",
    "    embl = open(embl_path, 'r').read()\n",
    "    embl_records = embl.split('//')\n",
    "    embl_dict = dict()\n",
    "    for record in embl_records[:-1]:\n",
    "\n",
    "        embl_lines = record.split('\\n')\n",
    "        seq_id = None\n",
    "        description = None\n",
    "        seq_start = None\n",
    "        for ind, line in enumerate(embl_lines):\n",
    "            if line[:2] == 'ID':\n",
    "                seq_id = line[2:].strip()\n",
    "            elif line[:2] == 'DE':\n",
    "                description = line[2:].strip()\n",
    "            elif line[:2] == 'SQ':\n",
    "                seq_start = ind + 1\n",
    "                break\n",
    "        sequence = ''.join([line for line in embl_lines[seq_start:]]).replace(' ', '')\n",
    "        sequence = re.sub('[0-9]+', '', sequence).upper()\n",
    "        embl_dict[seq_id] = {'descrption': description, 'sequence': sequence}\n",
    "    embl_df = pd.DataFrame(embl_dict).T.reset_index()\n",
    "    embl_df.rename(columns={'index': 'Seq_ID'}, inplace=True)\n",
    "    if write is True:\n",
    "        embl_df.to_csv('embl_info.csv')\n",
    "    return embl_df\n",
    "        \n",
    "\n",
    "check_all_seqs('Test.txt')\n",
    "embl_df = parse_embl('EMBL_records.txt')\n",
    "\n",
    "embl_df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using a Custom wrapper through the sh library \n",
    "### Only available in unix-like systems"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-c09e4c884765>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0mblast_output\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'mitochondrial.tab'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m \u001b[0mblastdb_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'/home/pradeep/blastdb/nt'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m \u001b[0mblastn_wrapper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfasta_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mblast_output\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mblastdb_path\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-4-c09e4c884765>\u001b[0m in \u001b[0;36mblastn_wrapper\u001b[0;34m(query, out, db, task, perc_identity, qcov_hsp_perc, max_target_seqs, threads, buffer, word_size)\u001b[0m\n\u001b[1;32m     26\u001b[0m                \u001b[0;34m'-qcov_hsp_perc'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mqcov_hsp_perc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-dust'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'no'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-max_target_seqs'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmax_target_seqs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m                \u001b[0;34m'-out'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-query'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mquery\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-outfmt'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutfmt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-word_size'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mword_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-gapopen'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'-gapextend'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 28\u001b[0;31m             '-evalue', 10)\n\u001b[0m\u001b[1;32m     29\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[0mfasta_path\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'mitochondrial.fasta'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dtra/lib/python3.7/site-packages/sh-1.14.1-py3.7.egg/sh.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1518\u001b[0m             \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"wb\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1519\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1520\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mRunningCommand\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcall_args\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstdin\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1521\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1522\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dtra/lib/python3.7/site-packages/sh-1.14.1-py3.7.egg/sh.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, cmd, call_args, stdin, stdout, stderr)\u001b[0m\n\u001b[1;32m    782\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    783\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mshould_wait\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 784\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    785\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    786\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dtra/lib/python3.7/site-packages/sh-1.14.1-py3.7.egg/sh.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    829\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    830\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 831\u001b[0;31m                 \u001b[0mexit_code\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    832\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_waited_until_completion\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    833\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dtra/lib/python3.7/site-packages/sh-1.14.1-py3.7.egg/sh.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   2461\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexit_code\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2462\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlog\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdebug\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"exit code not set, waiting on pid\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2463\u001b[0;31m                 \u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mexit_code\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mno_interrupt\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwaitpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpid\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m  \u001b[0;31m# blocks\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2464\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexit_code\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mhandle_process_exit_code\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mexit_code\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2465\u001b[0m                 \u001b[0mwitnessed_end\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/dtra/lib/python3.7/site-packages/sh-1.14.1-py3.7.egg/sh.py\u001b[0m in \u001b[0;36mno_interrupt\u001b[0;34m(syscall, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1760\u001b[0m     \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1761\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1762\u001b[0;31m             \u001b[0mret\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msyscall\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1763\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mOSError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1764\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merrno\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0merrno\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mEINTR\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "\n",
    "from sh import blastn\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline as bioblastn\n",
    "from Bio import SearchIO\n",
    "import pandas as pd\n",
    "\n",
    "\n",
    "def blastn_wrapper(query, out, db='protein', task='megablast', perc_identity=92, qcov_hsp_perc=100,\n",
    "                           max_target_seqs=15, threads=60, buffer=False, word_size=15):\n",
    "\n",
    "    \"\"\"\n",
    "    Input:\n",
    "    query: str, path to fasta file of universal regions\n",
    "    out: str, path to output blast tab file\n",
    "    perc_identity: int, percentage identical to query entries\n",
    "    qcov_hsp_perc: int, percentage of query entry covered by hits\n",
    "    max_target_seqs: int, number of hits to show in tab file\n",
    "    Used for specificity analysis of final universal regions.\n",
    "    \"\"\"\n",
    "\n",
    "    if not db:\n",
    "        db = 'protein'\n",
    "\n",
    "    outfmt = \"6\" # qseqid sseqid pident qcovhsp length mismatch gapopen qstart qend sstart send evalue bitscore\"\n",
    "\n",
    "    blastn('-db', db, '-perc_identity', perc_identity, '-task', task,\n",
    "               '-qcov_hsp_perc', qcov_hsp_perc, '-dust', 'no', '-max_target_seqs', max_target_seqs,\n",
    "               '-out', out, '-query', query, '-outfmt', outfmt, '-word_size', word_size, '-gapopen', 5, '-gapextend', 2,\n",
    "            '-evalue', 10)\n",
    "\n",
    "fasta_path = 'mitochondrial.fasta'\n",
    "blast_output = 'mitochondrial.tab'\n",
    "blastdb_path = '/home/pradeep/blastdb/nt'\n",
    "blastn_wrapper(fasta_path, blast_output, db = blastdb_path)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using Biopythons Blast Wrapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn -out mitochondrial.txt -query mitochondrial.fasta -db nt -remote -task megablast\n"
     ]
    }
   ],
   "source": [
    "blast_txt = 'mitochondrial.txt'\n",
    "blast_cline = bioblastn(query=fasta_path, out=blast_txt, task='megablast', remote=True, db='nt',)\n",
    "stdout, stderr = blast_cline()\n",
    "print(blast_cline)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing using biopython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Program: blastn (2.10.1+)\n",
      "  Query: NM_001319878.1 (788)\n",
      "         Camelus dromedarius superoxide dismutase 2 (SOD2),mRNA; nuclear gene...\n",
      " Target: Nucleotide collection (nt)\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  NM_001319878.1  Camelus dromedarius superoxide dismutas...\n",
      "            1      1  XM_032485290.1  PREDICTED: Camelus ferus superoxide dis...\n",
      "            2      1  XM_010965766.1  PREDICTED: Camelus bactrianus superoxid...\n",
      "            3      1  XM_015246572.2  PREDICTED: Vicugna pacos superoxide dis...\n",
      "Program: blastn (2.10.1+)\n",
      "  Query: XM_003872801.1 (490)\n",
      "         Leishmania mexicana MHOM/GT/2001/U1103 mitochondrialpartial mRNA\n",
      " Target: Nucleotide collection (nt)\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  XM_003872801.1  Leishmania mexicana MHOM/GT/2001/U1103 ...\n",
      "            1      1  FR799563.1  Leishmania mexicana MHOM/GT/2001/U1103 comp...\n",
      "            2      1  CP040138.1  Leishmania amazonensis strain UA301 chromos...\n",
      "            3      1  FR796406.1  Leishmania major strain Friedlin complete g...\n",
      "            4      1  XM_001681290.1  Leishmania major strain Friedlin putati...\n",
      "Program: blastn (2.10.1+)\n",
      "  Query: XM_001463625.1 (560)\n",
      "         Leishmania infantum JPCM5 mitochondrial partial mRNA\n",
      " Target: Nucleotide collection (nt)\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  CP022625.1  Leishmania donovani strain pasteur chromoso...\n",
      "            1      1  CP019517.1  Leishmania donovani strain MHOM/IN/1983/AG8...\n",
      "            2      1  CP018576.1  Leishmania donovani strain MHOM/IN/1983/AG8...\n",
      "            3      1  XM_003858804.1  Leishmania donovani DEAD/DEAH box helic...\n",
      "            4      1  FR799597.2  Leishmania donovani BPK282A1 complete genom...\n",
      "            5      1  CP048173.1  Leishmania chagasi strain MCER/BR/1981/M644...\n",
      "            6      1  LR812943.1  Leishmania infantum genome assembly, chromo...\n",
      "            7      1  LR812630.1  Leishmania donovani genome assembly, chromo...\n",
      "            8      1  FR796442.1  Leishmania infantum JPCM5 genome chromosome 10\n",
      "            9      1  XM_001463625.1  Leishmania infantum JPCM5 mitochondrial...\n",
      "           10      1  CP029509.1  Leishmania donovani strain LdCL chromosome ...\n",
      "           11      1  CP048137.1  Leishmania chagasi strain MHOM/HD/2017/M325...\n",
      "           12      1  CP027808.1  Leishmania infantum strain TR01 isolate Lin...\n",
      "           13      1  FR796406.1  Leishmania major strain Friedlin complete g...\n",
      "Program: blastn (2.10.1+)\n",
      "  Query: XM_003392155.1 (490)\n",
      "         Leishmania infantum JPCM5 mitochondrial partial mRNA\n",
      " Target: Nucleotide collection (nt)\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  CP029502.1  Leishmania donovani strain LdCL chromosome ...\n",
      "            1      1  CP022618.1  Leishmania donovani strain pasteur chromoso...\n",
      "            2      1  CP019546.1  Leishmania donovani strain MHOM/IN/1983/AG8...\n",
      "            3      1  CP018605.1  Leishmania donovani strain MHOM/IN/1983/AG8...\n",
      "            4      1  XM_003857977.1  Leishmania donovani MP99, putative (LDB...\n",
      "            5      1  CP048202.1  Leishmania chagasi strain MCER/BR/1981/M644...\n",
      "            6      1  CP048166.1  Leishmania chagasi strain MHOM/HD/2017/M325...\n",
      "            7      1  LR812623.1  Leishmania donovani genome assembly, chromo...\n",
      "            8      1  LR812936.1  Leishmania infantum genome assembly, chromo...\n",
      "            9      1  FR799590.1  Leishmania donovani BPK282A1 complete genom...\n",
      "           10      1  FR796435.1  Leishmania infantum JPCM5 genome chromosome 3\n",
      "           11      1  XM_003392155.1  Leishmania infantum JPCM5 mitochondrial...\n",
      "           12      1  CP027829.1  Leishmania infantum strain TR01 isolate Lin...\n",
      "           13      1  FR796399.1  Leishmania major strain Friedlin complete g...\n",
      "           14      1  XM_003721690.1  Leishmania major strain Friedlin hypoth...\n",
      "Program: blastn (2.10.1+)\n",
      "  Query: NM_138591.2 (350)\n",
      "         Mus musculus G elongation factor, mitochondrial 1(Gfm1), mRNA; nucle...\n",
      " Target: Nucleotide collection (nt)\n",
      "   Hits: ----  -----  ----------------------------------------------------------\n",
      "            #  # HSP  ID + description\n",
      "         ----  -----  ----------------------------------------------------------\n",
      "            0      1  NM_138591.2  Mus musculus G elongation factor, mitochon...\n",
      "            1      1  AK033214.1  Mus musculus 15 days embryo male testis cDN...\n",
      "            2      1  BC013093.1  Mus musculus G elongation factor, mitochond...\n",
      "            3      1  XM_021158645.1  PREDICTED: Mus caroli G elongation fact...\n",
      "            4      1  XM_021195693.1  PREDICTED: Mus pahari G elongation fact...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/alarsen/miniconda3/envs/dtra/lib/python3.7/site-packages/Bio/SearchIO/_legacy/NCBIStandalone.py:93: BiopythonWarning: Parsing BLAST plain text output file is not a well supported functionality anymore. Consider generating your BLAST output for parsing as XML or tabular format instead.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "bltab = SearchIO.parse('mitochondrial.txt', 'blast-text')\n",
    "for line in bltab:\n",
    "    print(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Parsing using Pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query</th>\n",
       "      <th>subject</th>\n",
       "      <th>pc_identity</th>\n",
       "      <th>aln_length</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>gaps_opened</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>subject_start</th>\n",
       "      <th>subject_end</th>\n",
       "      <th>e_value</th>\n",
       "      <th>bitscore</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>CP018576.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>43332</td>\n",
       "      <td>43891</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>XM_003858804.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>FR799597.2</td>\n",
       "      <td>100.000</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>43331</td>\n",
       "      <td>43890</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>CP048173.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>44827</td>\n",
       "      <td>45386</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>LR812943.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>56335</td>\n",
       "      <td>56894</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>LR812630.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>56281</td>\n",
       "      <td>56840</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>FR796442.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>44829</td>\n",
       "      <td>45388</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>560</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>CP029509.1</td>\n",
       "      <td>99.821</td>\n",
       "      <td>560</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>56327</td>\n",
       "      <td>56886</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>CP048137.1</td>\n",
       "      <td>99.821</td>\n",
       "      <td>560</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>44821</td>\n",
       "      <td>45380</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>CP027808.1</td>\n",
       "      <td>99.467</td>\n",
       "      <td>563</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>44947</td>\n",
       "      <td>45509</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1046</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>XM_001463625.1</td>\n",
       "      <td>FR796406.1</td>\n",
       "      <td>92.143</td>\n",
       "      <td>560</td>\n",
       "      <td>41</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>560</td>\n",
       "      <td>57102</td>\n",
       "      <td>57658</td>\n",
       "      <td>0.0</td>\n",
       "      <td>813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>CP029502.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>227425</td>\n",
       "      <td>227914</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>CP022618.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>233950</td>\n",
       "      <td>234439</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>CP019546.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>228755</td>\n",
       "      <td>229244</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>CP018605.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>228822</td>\n",
       "      <td>229311</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>XM_003857977.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>CP048202.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>230041</td>\n",
       "      <td>230530</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>CP048166.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>229964</td>\n",
       "      <td>230453</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>LR812623.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>228231</td>\n",
       "      <td>228720</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>LR812936.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>230112</td>\n",
       "      <td>230601</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>FR799590.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>228818</td>\n",
       "      <td>229307</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>FR796435.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>230047</td>\n",
       "      <td>230536</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>100.000</td>\n",
       "      <td>490</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>CP027829.1</td>\n",
       "      <td>99.391</td>\n",
       "      <td>493</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>230939</td>\n",
       "      <td>231431</td>\n",
       "      <td>0.0</td>\n",
       "      <td>921</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>FR796399.1</td>\n",
       "      <td>97.755</td>\n",
       "      <td>490</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>228696</td>\n",
       "      <td>229185</td>\n",
       "      <td>0.0</td>\n",
       "      <td>879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>XM_003392155.1</td>\n",
       "      <td>XM_003721690.1</td>\n",
       "      <td>97.755</td>\n",
       "      <td>490</td>\n",
       "      <td>11</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>1</td>\n",
       "      <td>490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NM_138591.2</td>\n",
       "      <td>NM_138591.2</td>\n",
       "      <td>100.000</td>\n",
       "      <td>350</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>350</td>\n",
       "      <td>1</td>\n",
       "      <td>350</td>\n",
       "      <td>0.0</td>\n",
       "      <td>673</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             query         subject  pc_identity  aln_length  mismatches  \\\n",
       "0   XM_001463625.1      CP018576.1      100.000         560           0   \n",
       "1   XM_001463625.1  XM_003858804.1      100.000         560           0   \n",
       "2   XM_001463625.1      FR799597.2      100.000         560           0   \n",
       "3   XM_001463625.1      CP048173.1      100.000         560           0   \n",
       "4   XM_001463625.1      LR812943.1      100.000         560           0   \n",
       "5   XM_001463625.1      LR812630.1      100.000         560           0   \n",
       "6   XM_001463625.1      FR796442.1      100.000         560           0   \n",
       "7   XM_001463625.1  XM_001463625.1      100.000         560           0   \n",
       "8   XM_001463625.1      CP029509.1       99.821         560           1   \n",
       "9   XM_001463625.1      CP048137.1       99.821         560           1   \n",
       "10  XM_001463625.1      CP027808.1       99.467         563           0   \n",
       "11  XM_001463625.1      FR796406.1       92.143         560          41   \n",
       "12  XM_003392155.1      CP029502.1      100.000         490           0   \n",
       "13  XM_003392155.1      CP022618.1      100.000         490           0   \n",
       "14  XM_003392155.1      CP019546.1      100.000         490           0   \n",
       "15  XM_003392155.1      CP018605.1      100.000         490           0   \n",
       "16  XM_003392155.1  XM_003857977.1      100.000         490           0   \n",
       "17  XM_003392155.1      CP048202.1      100.000         490           0   \n",
       "18  XM_003392155.1      CP048166.1      100.000         490           0   \n",
       "19  XM_003392155.1      LR812623.1      100.000         490           0   \n",
       "20  XM_003392155.1      LR812936.1      100.000         490           0   \n",
       "21  XM_003392155.1      FR799590.1      100.000         490           0   \n",
       "22  XM_003392155.1      FR796435.1      100.000         490           0   \n",
       "23  XM_003392155.1  XM_003392155.1      100.000         490           0   \n",
       "24  XM_003392155.1      CP027829.1       99.391         493           0   \n",
       "25  XM_003392155.1      FR796399.1       97.755         490          11   \n",
       "26  XM_003392155.1  XM_003721690.1       97.755         490          11   \n",
       "27     NM_138591.2     NM_138591.2      100.000         350           0   \n",
       "\n",
       "    gaps_opened  query_start  query_end  subject_start  subject_end  e_value  \\\n",
       "0             0            1        560          43332        43891      0.0   \n",
       "1             0            1        560              1          560      0.0   \n",
       "2             0            1        560          43331        43890      0.0   \n",
       "3             0            1        560          44827        45386      0.0   \n",
       "4             0            1        560          56335        56894      0.0   \n",
       "5             0            1        560          56281        56840      0.0   \n",
       "6             0            1        560          44829        45388      0.0   \n",
       "7             0            1        560              1          560      0.0   \n",
       "8             0            1        560          56327        56886      0.0   \n",
       "9             0            1        560          44821        45380      0.0   \n",
       "10            2            1        560          44947        45509      0.0   \n",
       "11            1            1        560          57102        57658      0.0   \n",
       "12            0            1        490         227425       227914      0.0   \n",
       "13            0            1        490         233950       234439      0.0   \n",
       "14            0            1        490         228755       229244      0.0   \n",
       "15            0            1        490         228822       229311      0.0   \n",
       "16            0            1        490              1          490      0.0   \n",
       "17            0            1        490         230041       230530      0.0   \n",
       "18            0            1        490         229964       230453      0.0   \n",
       "19            0            1        490         228231       228720      0.0   \n",
       "20            0            1        490         230112       230601      0.0   \n",
       "21            0            1        490         228818       229307      0.0   \n",
       "22            0            1        490         230047       230536      0.0   \n",
       "23            0            1        490              1          490      0.0   \n",
       "24            1            1        490         230939       231431      0.0   \n",
       "25            0            1        490         228696       229185      0.0   \n",
       "26            0            1        490              1          490      0.0   \n",
       "27            0            1        350              1          350      0.0   \n",
       "\n",
       "    bitscore  \n",
       "0       1077  \n",
       "1       1077  \n",
       "2       1077  \n",
       "3       1077  \n",
       "4       1077  \n",
       "5       1077  \n",
       "6       1077  \n",
       "7       1077  \n",
       "8       1071  \n",
       "9       1071  \n",
       "10      1046  \n",
       "11       813  \n",
       "12       942  \n",
       "13       942  \n",
       "14       942  \n",
       "15       942  \n",
       "16       942  \n",
       "17       942  \n",
       "18       942  \n",
       "19       942  \n",
       "20       942  \n",
       "21       942  \n",
       "22       942  \n",
       "23       942  \n",
       "24       921  \n",
       "25       879  \n",
       "26       879  \n",
       "27       673  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "headers = ['query', 'subject', 'pc_identity', 'aln_length', 'mismatches', 'gaps_opened',\n",
    "                   'query_start', 'query_end', 'subject_start', 'subject_end', 'e_value', 'bitscore']\n",
    "df = pd.read_csv('mitochondrial.tab', '\\t', skiprows=7, header=None)\n",
    "df.columns = headers\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
