{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter local Fasta:  BRCA1.faa\n",
      "Enter Motif:  QKVNE\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Fasta header: NP_009225.1 BRCA1 [organism=Homo sapiens] [GeneID=672] [isoform=1]\n",
      "\n",
      "Found motif: QKVNE, 1 time(s).\n",
      "Motif first found at positions 379 - 384\n",
      "\n",
      "\n",
      "Fasta header: NP_009228.2 BRCA1 [organism=Homo sapiens] [GeneID=672] [isoform=3]\n",
      "\n",
      "Found motif: QKVNE, 1 time(s).\n",
      "Motif first found at positions 332 - 337\n",
      "\n",
      "\n",
      "Fasta header: NP_009229.2 BRCA1 [organism=Homo sapiens] [GeneID=672] [isoform=4]\n",
      "\n",
      "Unable to find QKVNE\n",
      "\n",
      "\n",
      "Fasta header: NP_009230.2 BRCA1 [organism=Homo sapiens] [GeneID=672] [isoform=5]\n",
      "\n",
      "Unable to find QKVNE\n",
      "\n",
      "\n",
      "Fasta header: NP_009231.2 BRCA1 [organism=Homo sapiens] [GeneID=672] [isoform=2]\n",
      "\n",
      "Found motif: QKVNE, 1 time(s).\n",
      "Motif first found at positions 379 - 384\n",
      "\n",
      "\n",
      "Motif discovered 3 time(s) across 5 records.\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import os\n",
    "\n",
    "def load_fasta(fasta_path):\n",
    "    \"\"\"Without Using Biopython\"\"\"\n",
    "    fasta = open(fasta_path, 'r').read()\n",
    "    record_dict = dict()\n",
    "    records = fasta.split('>')\n",
    "    for rec in records:\n",
    "        if rec:\n",
    "            rec_lines = rec.split('\\n')\n",
    "            header = rec_lines[0]\n",
    "            seq = ''.join(rec_lines[1:]).replace(' ', '')\n",
    "            record_dict[header] = seq\n",
    "    return record_dict\n",
    "\n",
    "def check_seq_for_motif(seq, motif):\n",
    "    \"\"\"\n",
    "    Check seq for motif\n",
    "    Input:\n",
    "    seq: str, sequence\n",
    "    motif: str, motif to search seq for\n",
    "    \"\"\"\n",
    "    r_all = re.findall(motif, seq)\n",
    "    r_first = re.search(motif, seq)\n",
    "    motif_count = len(r_all)\n",
    "    if r_first:\n",
    "        start, end = r_first.span()\n",
    "        print(f'Found motif: {motif}, {motif_count} time(s).')\n",
    "        print(f'Motif first found at positions {start} - {end}\\n\\n')\n",
    "    else:\n",
    "        print(f'Unable to find {motif}\\n\\n')\n",
    "        \n",
    "    return motif_count\n",
    "        \n",
    "        \n",
    "def check_all_seqs(fasta_path, motif):\n",
    "    \"\"\"\n",
    "    Input:\n",
    "    fasta_path: str, path to fasta file\n",
    "    motif: str, motif to search for within fasta\n",
    "    \n",
    "    \"\"\"\n",
    "    if os.path.exists(fasta_path):\n",
    "        record_dict = load_fasta(fasta_path)\n",
    "    else:\n",
    "        print(f'Fasta path {fasta_path} does not exist.')\n",
    "        raise OSError\n",
    "    total_count = 0\n",
    "    N_records = 0\n",
    "    for header in record_dict:\n",
    "        N_records += 1\n",
    "        print(f'Fasta header: {header}\\n')\n",
    "        tcount = check_seq_for_motif(record_dict[header], motif)\n",
    "        total_count += tcount\n",
    "    print(f'Motif discovered {total_count} time(s) across {N_records} records.')\n",
    "\n",
    "#check_all_seqs('BRCA1.faa', 'QKVNE')\n",
    "\n",
    "try_count = 0\n",
    "while try_count < 3:\n",
    "    try:\n",
    "        fasta_path = input('Enter local Fasta: ')\n",
    "        motif = input('Enter Motif: ')\n",
    "        print('\\n')\n",
    "        check_all_seqs(fasta_path, motif)\n",
    "        break\n",
    "    except Exception:\n",
    "        print('Issue with fasta input, check fasta.')\n",
    "    try_count += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NP_009225.1\n",
      "NP_009228.2\n",
      "NP_009229.2\n",
      "NP_009230.2\n",
      "NP_009231.2\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "def load_fasta(fasta_path):\n",
    "    \"\"\"Without Using Biopython\"\"\"\n",
    "    fasta = open(fasta_path, 'r').read()\n",
    "    record_dict = dict()\n",
    "    records = fasta.split('>')\n",
    "    for rec in records:\n",
    "        if rec:\n",
    "            rec_lines = rec.split('\\n')\n",
    "            header = rec_lines[0]\n",
    "            seq = ''.join(rec_lines[1:]).replace(' ', '')\n",
    "            record_dict[header] = seq\n",
    "    return record_dict\n",
    "\n",
    "records = load_fasta('BRCA1.faa')\n",
    "for header in records:\n",
    "    r = re.search('\\w+\\.\\w', header)\n",
    "    print(r.group())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(RNAi) ends at position 50\n",
      "\n",
      "RNAi ends at position 137\n",
      "\n",
      "(dsRNA) ends at position 212\n",
      "\n",
      "dsRNA ends at position 331\n",
      "\n",
      "(ssRNAs) ends at position 375\n",
      "\n",
      "RNAs ends at position 380\n",
      "\n",
      "mRNAs ends at position 431\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "text =  \"\"\"Several rapidly developing RNA interference (RNAi)\n",
    "methodologies hold the promise to selectively inhibit gene expression in\n",
    "mammals. RNAi is an innate cellular process activated when a\n",
    "double-stranded RNA (dsRNA) molecule of greater than 19 duplex\n",
    "nucleotides enters the cell, causing the degradation of not only the\n",
    "invading dsRNA molecule, but also single-stranded (ssRNAs) RNAs of\n",
    "identical sequences, including endogenous mRNAs.\"\"\"\n",
    "\n",
    "results = re.finditer('[\\w\\(\\)]*RNA[\\w\\(\\)]*', text)\n",
    "match_list = []\n",
    "for r in results:\n",
    "    string = r.group()\n",
    "    end = r.span()[1]\n",
    "    if string == 'RNA':\n",
    "        continue\n",
    "    if string not in match_list:\n",
    "        match_list.append(string)\n",
    "    else:\n",
    "        continue\n",
    "    format_line = f'{string} ends at position {end}\\n'\n",
    "    print(format_line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 5 records\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'MDLSALRVEEVQNVINAMQKILECPICLELIKEPVSTKCDHIFCKFCMLKLLNQKKGPSQCPLCKNDITKRSLQESTRFSQLVEELLKIICAFQLDTGLE'"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def load_fasta(fasta_path):\n",
    "    \"\"\"Without Using Biopython\"\"\"\n",
    "    fasta = open(fasta_path, 'r').read()\n",
    "    record_dict = dict()\n",
    "    records = fasta.split('>')\n",
    "    for rec in records:\n",
    "        if rec:\n",
    "            rec_lines = rec.split('\\n')\n",
    "            header = rec_lines[0].split(' ')[0]\n",
    "            seq = ''.join(rec_lines[1:]).replace(' ', '')\n",
    "            record_dict[header] = seq[:100]\n",
    "    print(f'Loaded {len(record_dict)} records')\n",
    "    return record_dict\n",
    "\n",
    "rec_dict = load_fasta('BRCA1.faa')\n",
    "rec_dict#['NP_009225.1']\n",
    "\n",
    "print(rec_dict['NP_009225.1'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Enter in fasta file path:  seq1.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Header:  seq1\n",
      "AWTHRSISRRGTFLEDVEGLEQLPRAVRKRRLQCGR_IAS_MERLG_RPGEKCLSPRGFTTSQELYKTNRTSSFLRCGQCSLIPLLQGGI_ATKSMLRFAQKGRFSIYFMSNRLRNYLCREVSGTNKHNRGSV\n",
      "AA seq:  None\n"
     ]
    }
   ],
   "source": [
    "def load_fasta(fasta_path):\n",
    "    \"\"\"Without Using Biopython\"\"\"\n",
    "    fasta = open(fasta_path, 'r').read()\n",
    "    record_dict = dict()\n",
    "    records = fasta.split('>')\n",
    "    for rec in records:\n",
    "        if rec:\n",
    "            rec_lines = rec.split('\\n')\n",
    "            header = rec_lines[0]\n",
    "            seq = ''.join(rec_lines[1:]).replace(' ', '')\n",
    "            record_dict[header] = seq\n",
    "    return record_dict\n",
    "\n",
    "gencode = {\n",
    "'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
    "'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
    "'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
    "'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
    "'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
    "'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
    "'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
    "'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
    "'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
    "'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
    "'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
    "'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
    "'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
    "'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
    "'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W'}\n",
    "\n",
    "def get_aa(seq):\n",
    "    seq_len = len(seq)\n",
    "    aa_seq = ''\n",
    "    for ind in range(0, seq_len, 3):\n",
    "        end = ind + 3\n",
    "        codon = seq[ind:end]\n",
    "        if len(codon) != 3:\n",
    "            continue\n",
    "        aa = gencode[codon]\n",
    "        aa_seq += aa\n",
    "    print(aa_seq)\n",
    "\n",
    "inpu = input('Enter in fasta file path: ')\n",
    "records = load_fasta(inpu)\n",
    "for header in records:\n",
    "    print('Header: ', header)\n",
    "    aa = get_aa(records[header])\n",
    "    print('AA seq: ', aa)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1140, 1146)\n",
      "0 1143\n",
      "(1625, 1631)\n",
      "1143 1628\n",
      "Fragment lengths:  [1143, 485, 383]\n",
      "Average sequence length:  670.3333333333334\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "dna = open('dna.txt', 'r').read().strip('\\n')\n",
    "\n",
    "re_pattern = 'A[ATGC]TAAT'\n",
    "\n",
    "results = re.finditer(re_pattern, dna)\n",
    "seqs = []\n",
    "\n",
    "start = 0\n",
    "for r in results:\n",
    "    end = r.span()[1] - 3\n",
    "    seq = dna[start:end]\n",
    "    seqs.append(seq)\n",
    "    start = end\n",
    "seq = dna[start:-1]\n",
    "seqs.append(seq)\n",
    "seq_lens = [len(seq) for seq in seqs]\n",
    "print('Fragment lengths: ', seq_lens)\n",
    "print('Average sequence length: ', sum(seq_lens) / len(seq_lens))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AaaI (XmaIII)': 'C^GGCCG',\n",
       " 'AacI (BamHI)': 'GGATCC',\n",
       " 'AaeI (BamHI)': 'GGATCC',\n",
       " 'AagI (ClaI)': 'AT^CGAT',\n",
       " 'AaqI (ApaLI)': 'GTGCAC',\n",
       " 'AarI': '^NNNNNNNNGCAGGTG',\n",
       " 'AatI (StuI)': 'AGG^CCT',\n",
       " 'AatII': 'GACGT^C',\n",
       " 'AauI (Bsp1407I)': 'T^GTACA',\n",
       " 'AbaI (BclI)': 'T^GATCA',\n",
       " 'AbeI (BbvCI)': 'GC^TGAGG',\n",
       " 'AbrI (XhoI)': 'C^TCGAG',\n",
       " 'AcaI (AsuII)': 'TTCGAA',\n",
       " 'AcaII (BamHI)': 'GGATCC',\n",
       " 'AcaIII (MstI)': 'TGCGCA',\n",
       " 'AcaIV (HaeIII)': 'GGCC',\n",
       " 'AccI': 'GT^MKAC',\n",
       " 'AccII (FnuDII)': 'CG^CG',\n",
       " 'AccIII (BspMII)': 'T^CCGGA',\n",
       " 'Acc16I (MstI)': 'TGC^GCA',\n",
       " 'Acc36I (BspMI)': '^NNNNNNNNGCAGGT',\n",
       " 'Acc38I (EcoRII)': 'CCWGG',\n",
       " 'Acc65I (KpnI)': 'G^GTACC',\n",
       " 'Acc113I (ScaI)': 'AGT^ACT',\n",
       " 'AccB1I (HgiCI)': 'G^GYRCC',\n",
       " 'AccB2I (HaeII)': 'RGCGC^Y',\n",
       " 'AccB7I (PflMI)': 'CCANNNN^NTGG',\n",
       " 'AccBSI (BsrBI)': 'GAG^CGG',\n",
       " 'AccEBI (BamHI)': 'G^GATCC',\n",
       " 'AceI (TseI)': 'G^CWGC',\n",
       " 'AceII (NheI)': 'GCTAG^C',\n",
       " 'AceIII': '^NNNNNNNNNNNGAGCTG',\n",
       " 'AciI': 'G^CGG',\n",
       " 'AclI': 'AA^CGTT',\n",
       " 'AclNI (SpeI)': 'A^CTAGT',\n",
       " 'AclWI (BinI)': 'GGATCNNNN^'}"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "enzy_dict = pd.read_csv('bionet.txt', sep='\\s{3,}', header=None, skiprows=9, engine='python', index_col=0).to_dict()[1]\n",
    "\n",
    "enzy_dict2 = dict()\n",
    "with open('bionet.txt', 'r') as bionet:\n",
    "    bionet = bionet.read()\n",
    "    results = re.findall('.+I.+', bionet)\n",
    "    for r in results:\n",
    "        key, val = re.split('\\s{3,}', r)\n",
    "        enzy_dict2[key] = val\n",
    "    \n",
    "\n",
    "enzy_dict2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
